core/templates/dev/head/forms/formBuilder.js

```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+      }
+    };
+  }
+]);
+
+// Dynamically generate CKEditor widgets for the rich text components.
+oppia.run([
+  '$timeout', '$compile', '$rootScope', 'rteHelperService', 'oppiaHtmlEscaper',
+  function($timeout, $compile, $rootScope, rteHelperService, oppiaHtmlEscaper) {
+    var _RICH_TEXT_COMPONENTS = rteHelperService.getRichTextComponents();
+    _RICH_TEXT_COMPONENTS.forEach(function(componentDefn) {
+      // The name of the CKEditor widget corresponding to this component.
+      var ckName = 'oppia' + componentDefn.name;
+      // For some reason, frontend tests will error without this check.
+      if (CKEDITOR.plugins.registered[ckName] !== undefined) {
+        return;
+      }
+      var tagName = 'oppia-noninteractive-' + componentDefn.name;
+      var customizationArgSpecs = componentDefn.customizationArgSpecs;
+      var isInline = rteHelperService.isInlineComponent(componentDefn.name);
+      // Inline components will be wrapped in a span, while block components
+      // will be wrapped in a div.
+      if (isInline) {
+        var componentTemplate = '<span type="' + tagName + '">' +
+                                '<' + tagName + '></' + tagName + '>' +
+                                '</span>';
+      } else {
+        var componentTemplate = '<div class="component-container" type="' +
+                                tagName + '">' +
+                                '<' + tagName + '></' + tagName + '>' +
+                                '<div class="component-overlay"></div>' +
+                                '</div>';
+      }
+      CKEDITOR.plugins.add(ckName, {
+        init: function(editor) {
+          // Create the widget itself.
+          editor.widgets.add(ckName, {
+            button: componentDefn.tooltip,
+            inline: isInline,
+            template: componentTemplate,
+            edit: function(event) {
+              editor.fire('lockSnapshot', {
+                dontUpdate: true
+              });
+              // Prevent default action since we are using our own edit modal.
+              event.cancel();
+              // Save this for creating the widget later.
+              var container = this.wrapper.getParent(true);
+              var that = this;
+              var customizationArgs = {};
+              customizationArgSpecs.forEach(function(spec) {
+                customizationArgs[spec.name] = that.data[spec.name] ||
+                                               spec.default_value;
+              });
+
+              rteHelperService._openCustomizationModal(
+                customizationArgSpecs,
+                customizationArgs,
+                function(customizationArgsDict) {
+                  for (var arg in customizationArgsDict) {
+                    if (customizationArgsDict.hasOwnProperty(arg)) {
+                      that.setData(arg, customizationArgsDict[arg]);
+                    }
+                  }
+
+                  if (!that.isReady()) {
```

Comment:
Well... this is a bit confusing. When exactly is creation supposed to happen? In what circumstances do we need the guard here?